Include/internal/pycore_stackref.h

```diff
@@ -0,0 +1,189 @@
+    }
+#else
+    #define PyStackRef_StealObject(tagged) PyStackRef_Get(tagged)
+#endif
+
+static inline void
+_Py_untag_stack_borrowed(PyObject **dst, const _PyStackRef *src, size_t length) {
+    for (size_t i = 0; i < length; i++) {
+        dst[i] = PyStackRef_Get(src[i]);
+    }
+}
+
+static inline void
+_Py_untag_stack_steal(PyObject **dst, const _PyStackRef *src, size_t length) {
+    for (size_t i = 0; i < length; i++) {
+        dst[i] = PyStackRef_StealObject(src[i]);
+    }
+}
+
+
+#define PyStackRef_XSETREF(dst, src) \
+    do { \
+        _PyStackRef *_tmp_dst_ptr = _Py_CAST(_PyStackRef*, &(dst)); \
+        _PyStackRef _tmp_old_dst = (*_tmp_dst_ptr); \
+        *_tmp_dst_ptr = (src); \
+        PyStackRef_XDECREF(_tmp_old_dst); \
+    } while (0)
+
+#define PyStackRef_SETREF(dst, src) \
+    do { \
+        _PyStackRef *_tmp_dst_ptr = _Py_CAST(_PyStackRef*, &(dst)); \
+        _PyStackRef _tmp_old_dst = (*_tmp_dst_ptr); \
+        *_tmp_dst_ptr = (src); \
+        PyStackRef_DECREF(_tmp_old_dst); \
+    } while (0)
+
+#define PyStackRef_CLEAR(op) \
+    do { \
+        _PyStackRef *_tmp_op_ptr = _Py_CAST(_PyStackRef*, &(op)); \
+        _PyStackRef _tmp_old_op = (*_tmp_op_ptr); \
+        if (PyStackRef_Get(_tmp_old_op) != NULL) { \
+            *_tmp_op_ptr = PyStackRef_StealRef(_Py_NULL); \
+            PyStackRef_DECREF(_tmp_old_op); \
+        } \
+    } while (0)
+
+#if defined(Py_GIL_DISABLED)
+    static inline void
+    PyStackRef_DECREF(_PyStackRef tagged) {
+        if ((tagged.bits & Py_TAG_DEFERRED) == Py_TAG_DEFERRED) {
+            return;
+        }
+        Py_DECREF(PyStackRef_Get(tagged));
+    }
+#else
+    #define PyStackRef_DECREF(op) Py_DECREF(PyStackRef_Get(op))
+#endif
+
+#define PyStackRef_DECREF_OWNED(op) Py_DECREF(PyStackRef_Get(op));
+
+#if defined(Py_GIL_DISABLED)
+    static inline void
+    PyStackRef_INCREF(_PyStackRef tagged) {
+        if ((tagged.bits & Py_TAG_DEFERRED) == Py_TAG_DEFERRED) {
+            assert(_PyObject_HasDeferredRefcount(PyStackRef_Get(tagged)));
+            return;
+        }
+        Py_INCREF(PyStackRef_Get(tagged));
+    }
+#else
+    #define PyStackRef_INCREF(op) Py_INCREF(PyStackRef_Get(op))
+#endif
+
+static inline void
+PyStackRef_XDECREF(_PyStackRef op)
+{
+    if (op.bits != Py_STACKREF_NULL.bits) {
+        PyStackRef_DECREF(op);
+    }
+}
+
+static inline _PyStackRef
+PyStackRef_NewRef(_PyStackRef obj)
+{
+    PyStackRef_INCREF(obj);
+    return obj;
+}
+
+static inline _PyStackRef
+PyStackRef_XNewRef(_PyStackRef obj)
+{
+    if (obj.bits == Py_STACKREF_NULL.bits) {
+        return obj;
+    }
+    return PyStackRef_NewRef(obj);
+}
+
+// Converts a PyObject * to a PyStackRef, with a new reference
+#if defined(Py_GIL_DISABLED)
+    static inline _PyStackRef
+    _PyStackRef_NewRefDeferred(PyObject *obj) {
+        // Make sure we don't take an already tagged value.
+        assert(PyStackRef_Get(((_PyStackRef){.bits = ((uintptr_t)(obj))})) == obj);
```

Comment:
The comment makes me think that you want the following, but that's not what the code does:

```suggestion
        assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
```